### PR TITLE
Add batch manipulation functions.

### PR DESCRIPTION
**Pull request by chantera**
This PR adds following batch manipulation functions:
- `functions::batch::pick`
- `functions::batch::slice`
- `functions::batch::split`
- `functions::batch::concat`

### CODE REVIEW

**Comment by vbkaisetsu** [comment]
Related to #175

File: primitiv/device_ops/cuda/batch_slice.cu [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    const float *px, std::uint32_t size, float *py) {
+  const std::uint32_t i = IDX;
+  if (i < size) py[i] = px[i];
+}
+
+__global__ void batch_slice_bw_dev(
+    const float *pgy, std::uint32_t size, float *pgx) {
+  const std::uint32_t i = IDX;
+  if (i < size) {
+    ::atomicAdd(pgx + i, pgy[i]);
```
Comment (vbkaisetsu) [inline primitiv/device_ops/cuda/batch_slice.cu]: `pgx[i] += pgy[i]` ?

File: primitiv/device_ops/cuda16/batch_slice.cu [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  if (i < size) py[i] = px[i];
+}
+
+DECL_ATOMIC_OP(atomicHAdd, ::__fadd_rn);
+
+__global__ void batch_slice_bw_dev(
+    const half *pgy, std::uint32_t size, half *pgx) {
+  const std::uint32_t i = IDX;
+  if (i < size) {
+    ::atomicHAdd(pgx, i, ::__half2float(pgy[i]));
```
Comment (vbkaisetsu) [inline primitiv/device_ops/cuda16/batch_slice.cu]: ditto

File: primitiv/opencl_device_kernel.cl [hunk 3]
```diff
@@ -557,6 +581,23 @@ kernel void batch_sum_fw_kernel(
...
+  const unsigned ox = pi[bid_y * si] * sy;
+  const unsigned oy = bid_y * sy;
+  if (t < sy) atomic_add_float(pgx + ox + t, pgy[oy + t]);
+}
+
+kernel void batch_slice_bw_kernel(
+    const global float *pgy, const unsigned size,
+    global float *pgx, const unsigned shift) {
+  const unsigned i = get_global_id(0);
+  if (i < size) atomic_add_float(pgx + i + shift, pgy[i]);
```
Comment (vbkaisetsu) [inline primitiv/opencl_device_kernel.cl]: ditto

**Review by vbkaisetsu — COMMENTED** [review]
Thank you very much.
Please check some comments.

File: primitiv/basic_functions.h [hunk 4]
```diff
@@ -876,6 +876,86 @@ type_traits::Identity<Var> max_pool2d(
 
 namespace batch {
 
+/**
+ * Lookups subplanes according to the batch and addresses.
```
Comment (odashi) [inline primitiv/basic_functions.h]: (this line and others)
This function does not affect any dimensions so "subplane" looks strange.

File: primitiv/basic_functions.h [hunk 5]
```diff
@@ -876,6 +876,86 @@ type_traits::Identity<Var> max_pool2d(
...
+ * @return A list of `n` variables. Each variable is identical with:
+ *         ``batch::split(x, n)[i] == batch::slice(x, L(i), U(i))``,
+ *         where ``L(i) := i * x.shape().batch() / n``
+ *         and ``U(i) := (i + 1) * x.shape().batch() / n``.
+ * @throw primitiv::Error ``n`` can not divide ``s.shape().batch()`` without
+ *                        residue.
+ */
+template<typename Var>
+std::vector<type_traits::Identity<Var>> split(const Var &x, std::uint32_t n);
+
```
Comment (odashi) [inline primitiv/basic_functions.h:976]: (this line and others)
Add `@cond` and `@endcond` directives around all undocumented functions.
https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcond

File: primitiv/cuda16_device.h [hunk 6]
```diff
@@ -179,8 +179,16 @@ class CUDA16 : public Device {
   void sum_fw_impl(const Tensor &x, std::uint32_t dim, Tensor &y) override;
   void logsumexp_fw_impl(const Tensor &x, std::uint32_t dim, Tensor &y) override;
   void broadcast_fw_impl(const Tensor &x, std::uint32_t dim, std::uint32_t size, Tensor &y) override;
+
+  void batch_pick_fw_impl(const Tensor &x, const std::vector<std::uint32_t> &ids, Tensor &y) override;
+  void batch_slice_fw_impl(const Tensor &x, std::uint32_t offset, Tensor &y) override;
+  void batch_concat_fw_impl(
+      const std::vector<const Tensor *> &xs, Tensor &y) override;
```
Comment (odashi) [inline primitiv/cuda16_device.h]: (this line and others)
Why newline?

File: primitiv/shape_ops.h [hunk 7]
```diff
@@ -133,6 +133,37 @@ Shape pool2d(
     std::uint32_t padding0, std::uint32_t padding1,
     std::uint32_t stride0, std::uint32_t stride1);
 
+/**
+ * Calculates a picked shape.
```
Comment (odashi) [inline primitiv/shape_ops.h]: (this line and others)
Description also requires information that this function is a batch operation.

File: test/shape_ops_test.cc [hunk 8]
```diff
@@ -273,7 +274,7 @@ TEST_F(ShapeOpsTest, CheckInvalidPick) {
      {Shape({2, 2, 2}, 3), 0, {0, 1, 2}},
      {Shape({2, 2, 2}, 3), 1, {2}},
      {Shape({2, 2, 2}, 3), 2, {2}},
-     {Shape({2, 2, 2}, 3), 3, {1}},
+     {Shape({2, 2, 2}, 3), 3, {}},
```
Comment (odashi) [inline test/shape_ops_test.cc]: Restore original line. `{1}` is correct to evaluate the invalid operation: "pick the ID '1' from the axis with width 1."

File: test/shape_ops_test.cc [hunk 9]
```diff
@@ -560,5 +561,168 @@ TEST_F(ShapeOpsTest, CheckInvalidPool2D) {
...
+TEST_F(ShapeOpsTest, CheckBatchPick) {
+  struct TestCase {
+    Shape input;
+    vector<std::uint32_t> ids;
+    Shape expected;
+  };
+  const vector<TestCase> test_cases {
+    {Shape({2, 2, 2}, 3), {0, 0, 0}, Shape({2, 2, 2}, 3)},
+    {Shape({2, 2, 2}, 3), {1, 0, 1}, Shape({2, 2, 2}, 3)},
+    {Shape({2, 2, 2}, 3), {0}, Shape({2, 2, 2}, 1)},
```
Comment (odashi) [inline test/shape_ops_test.cc]: Batch size `1` can be removed.

File: test/shape_ops_test.cc [hunk 10]
```diff
@@ -560,5 +561,168 @@ TEST_F(ShapeOpsTest, CheckInvalidPool2D) {
...
+    {Shape({2, 2, 2}, 3), {2, 1, 1}, Shape({2, 2, 2}, 3)},
+    {Shape({2, 2, 2}, 3), {0, 1, 2, 1}, Shape({2, 2, 2}, 4)},
+    {{}, {0}, {}},
```
Comment (odashi) [inline test/shape_ops_test.cc]: Place simple test case(s) at the top of the list.

File: test/shape_ops_test.cc [hunk 11]
```diff
@@ -560,5 +561,168 @@ TEST_F(ShapeOpsTest, CheckInvalidPool2D) {
...
+    {{{}, {2}}, false},
+    {{{}, {1, 2}}, false},
+    {{{}, {1, 1, 2}}, false},
+    {{{}, {1, 1, 1, 2}}, false},
+    // volume is correct but dims is invalid
+    {{{2}, {2}}, true},
+    {{{2}, {1, 2}}, false},
+    {{{2}, {1, 1, 2}}, false},
+    {{{2}, {1, 1, 1, 2}}, false},
+    // only a shape is invalid
```
Comment (odashi) [inline test/shape_ops_test.cc]: a shape -> an axis

File: test/tensor_forward_test.cc [hunk 12]
```diff
@@ -2370,6 +2370,217 @@ TEST_F(TensorForwardTest, CheckInvalidBroadcast) {
...
+  for (Device *dev : devices) {
+    const Tensor x = dev->new_tensor_by_constant(Shape({2, 2}, 3), 0);
+    for (const TestCase &tc : test_cases) {
+      EXPECT_THROW(batch::pick(x, tc.ids), Error);
+    }
+  }
+}
+
+TEST_F(TensorForwardTest, CheckBatchSlice) {
+  vector<float> x_data = make_iota_vector(3 * 2 * 3, 0);
```
Comment (odashi) [inline test/tensor_forward_test.cc]: const.

File: test/tensor_forward_test.cc [hunk 13]
```diff
@@ -2370,6 +2370,217 @@ TEST_F(TensorForwardTest, CheckInvalidBroadcast) {
...
+  for (Device *dev : devices) {
+    const Tensor x = dev->new_tensor_by_constant(Shape({2, 2}, 3), 1);
+    for (const TestCase &tc : test_cases) {
+      EXPECT_THROW(batch::slice(x, tc.lower, tc.upper), Error);
+    }
+  }
+}
+
+TEST_F(TensorForwardTest, CheckBatchSplit) {
+  vector<float> x_data = make_iota_vector(3 * 3 * 4, 0);
```
Comment (odashi) [inline test/tensor_forward_test.cc]: const.

**Review by odashi — CHANGES_REQUESTED** [review]

File: test/shape_ops_test.cc (hunk 8 above)
Comment (chantera) [reply test/shape_ops_test.cc]: This PR includes your commit `https://github.com/primitiv/primitiv/commit/99987e73b51e636d6271b36843f3128df3c8c3a5`.
Could you check it?

Comment (odashi) [reply test/shape_ops_test.cc]: Ah, maybe this is my fault. I should fix this by myself.

File: primitiv/basic_functions.h [hunk 14]
```diff
@@ -877,18 +877,18 @@ type_traits::Identity<Var> max_pool2d(
 namespace batch {
 
 /**
- * Lookups subplanes according to the batch and addresses.
+ * Extracts elements according to the batch and addresses.
```
Comment (odashi) [inline primitiv/basic_functions.h]: "Element" is a general word and has less information for users (maybe not few users could not comprehend the behavior of the function by this document).

**Comment by chantera** [comment]
Fixed.

**Review by odashi — DISMISSED** [review]

**Review by odashi — APPROVED** [review]

**Pull request merged**